### ISSUE TITLE
refresh token logic got loop back to createRequestQueueInterceptor

### ISSUE THREAD

**Issue by imcyee** [Flyrell/axios-auth-refresh#120]
**Problem**: refresh token axios got loop back to createRequestQueueInterceptor, refreshing becomes stale

Hi~ thanks for the lib.
**Summary**: I have the same default axios running for normal request and refresh token, when it wants to run the refresh token, the request got queue to requestQueue, so it ended up stale. and the refreshing is never resolve.

**Here are the usage:** 

```
// Instantiate the interceptor (you can chain it as it returns the axios instance)
createAuthRefreshInterceptor(axios, refreshAuthLogic);

const refreshAuthLogic = (failedRequest: any) => {
  const asyncToPromise = async () => {
    // ...preprocessing
    const resp = await axios.post('/refreshUrl')
    const {accessToken} = resp.data
  // ...postprocessing
    return Promise.resolve(accessToken)
  }

  return asyncToPromise().then(async accessToken => {
    failedRequest.response.config.headers['Authorization'] = 'Bearer ' + accessToken;
    return Promise.resolve();
  })
};

```

**Current workaround:**
create another axios instance just for refreshing, so it wont get intercept

**Comment by Flyrell** [comment]
You need to skip the refreshing call from intercepting on failure.

```
const resp = await axios.post('/refreshUrl', { skipAuthRefresh: true });
const { accessToken } = resp.data;
```

**Comment by cYee992** [comment]
Thanks for the swift reply, i remembered I tried that this morning, try again still does not work. 
my loop intercept is actually at this line
https://github.com/Flyrell/axios-auth-refresh/blob/751c212686f7a84a7e1b8a24337234a069ba6fae/src/utils.ts#L89
```
 cache.requestQueueInterceptorId = instance.interceptors.request.use((request) => {
      return cache.refreshCall
          .catch(() => {
            throw new axios.Cancel('Request call failed');
          })
          .then(() => options.onRetry ? options.onRetry(request) : request);
    });
```

cache.refreshCall call my authRefreshLogic contains _axios_ call. and the request interceptor catch it and wait.

**Comment by Flyrell** [comment]
Do you use request interceptor to bind the new token to the stalled requests? 🙂  You can find more about it [here](https://github.com/Flyrell/axios-auth-refresh#request-interceptor).

**Comment by cYee992** [comment]
Thanks for that. Yup, I have read it. The problem i encountered is that the axios never makes out, but got intercepted in the **requestQueueInterceptor**, that awaiting **refreshing** to resolve. Which is the refreshing itself....

Flow: 401 -> ResponseIntercept -> createRequestQueue -> call refresh -> **axios** for requesting -> RequestIntercept intercept -> tell axios to wait for refreshing(But I am the refreshing)

**Comment by cYee992** [comment]
Will it because of the way i created my Promise authLogic? 

const refreshAuthLogic = (failedRequest: any) => {
  const asyncToPromise = async () => {
    // ...preprocessing
    const resp = await axios.post('/refreshUrl')
    const {accessToken} = resp.data
  // ...postprocessing
    return Promise.resolve(accessToken)
  }

  return asyncToPromise().then(async accessToken => {
    failedRequest.response.config.headers['Authorization'] = 'Bearer ' + accessToken;
    return Promise.resolve();
  })
};

**Comment by Flyrell** [comment]
You definitely need a `{ skipAuthRefresh: true }` flag on your request, then the interceptor and all should be working. Something like:

```typescript
const ACCESS_TOKEN_KEY = 'token';
function saveAccessToken(token) {
  localStorage.setItem(ACCESS_TOKEN_KEY, token);
}

function getAccessToken() {
  return localStorage.getItem(ACCESS_TOKEN_KEY);
}

const refreshAuthLogic = failedRequest => {
  // skipAuthRefresh will prevent the `axios-auth-refresh` from picking up the request when it fails (prevents the loop)
  const response = await axios.post('/refreshUrl', { skipAuthRefresh: true });

  // Store the token somewhere as every request from now on will need it (also the stalled requests will need it)
  saveAccessToken(response.data.accessToken);

  // Resolve the promise - signal for axios-auth-refresh that the refresh logic has completed
  return Promise.resolve();
};

// This is running right before each request is sent to the server
axios.interceptors.request.use(request => {
    // Makes sure each requests is using the latest accessToken available
    request.headers['Authorization'] = `Bearer ${getAccessToken()}`;
    return request;
});
```

**Comment by cYee992** [comment]
Hi~ Found the culpirit... my mistake of returning double promise....

  const asyncToPromise = async () => {
    // ...preprocessing
    const resp = await axios.post('/refreshUrl')
    const {accessToken} = resp.data
  // ...postprocessing
    **return Promise.resolve(accessToken)**
  }

**Comment by cYee992** [comment]
I found the issue. Actually it is not the double promise thingy. It is actually reproducible, by adding a **timeout** in your refreshAuthLogic before any axios request (to mimic async work, the axios will be stuck with cache.)


```
const refreshAuthLogic = async failedRequest => {
  
await new Promise((resolve)=>{
 setTimeout(()=>{
resolve()
}, 200)
})

// skipAuthRefresh will prevent the `axios-auth-refresh` from picking up the request when it fails (prevents the loop)
  const response = await axios.post('/refreshUrl', { skipAuthRefresh: true });

  // Store the token somewhere as every request from now on will need it (also the stalled requests will need it)
  saveAccessToken(response.data.accessToken);

  // Resolve the promise - signal for axios-auth-refresh that the refresh logic has completed
  return Promise.resolve();
};
```

**Comment by cYee992** [comment]
Can be fix with this

```
 cache.requestQueueInterceptorId = instance.interceptors.request.use((request) => {
   if (request.data.skipAuthRefresh) {
                return request
            }
      return cache.refreshCall
          .catch(() => {
            throw new axios.Cancel('Request call failed');
          })
          .then(() => options.onRetry ? options.onRetry(request) : request);
    });
```

**Comment by PhilippeCollin** [comment]
Isn't the **skipAuthRefresh** flag supposed to be passed in the request config? 
When using axios.post, that would be the third argument (and not the second) if I'm not mistaken.

**Comment by cYee992** [comment]
> Isn't the **skipAuthRefresh** flag supposed to be passed in the request config?
> When using axios.post, that would be the third argument (and not the second) if I'm not mistaken.

Thanks for pointing out let me try with the third argument instead

**Comment by cYee992** [comment]
> Isn't the **skipAuthRefresh** flag supposed to be passed in the request config?
> When using axios.post, that would be the third argument (and not the second) if I'm not mistaken.

Hi, philipper, I submitted another pr with skipAuthRefresh tag in config instead of data. The issue of this is because `createRequestQueueInterceptor` is created before calling the refreshing call. If there is any async logic is called before the axios call to refresh token. That refresh axios call will be looped back into the RequestQueueInterceptor. You can test this by adding a timeout in the refreshLogic before the axios.

**Comment by cYee992** [comment]
#121

**Comment by imcyee** [comment]
Sure thing.

**Comment by imcyee** [comment]
> Sorry for not answering. Had a busy end of the year but I'm trying to work on v3.1.0 now. Will take a look at that in your PR in upcoming days. Hope we can solve this

Btw, the summary is that 
createRequestQueueInterceptor -> intercept the supposed auth refresh -> unable to request 
Happens when you have other async task in your refreshLogic before the actual call.

Timeline:
axios -> server -> 403/401 -> client create a request interceptor -> refreshlogic -> any async logic in your refreshLogic -> refreshAxios -> caught in the request interceptor.

**Comment by SudoPlz** [comment]
I'm experiencing the same issue, I was wondering if there has been any progress on that.

Thank you

**Comment by darkbasic** [comment]
Same here. This is my code:

```ts
export const instance = axios.create({
  [...]
};

export async function retrieveToken(tokenType: 'access_token' | 'refresh_token') {
  [...]
}
export async function saveToken(tokenType: 'access_token' | 'refresh_token', token: string) {
  [...]
}

const refreshAuthLogic = async (failedRequest: any): Promise<void> => {
  console.log('Got 401, refreshing token', new Date());

  try {
    const config: AxiosAuthRefreshRequestConfig = {
      method: 'post',
      skipAuthRefresh: true,
      [...]
    };
    const {data: {access_token, refresh_token}}: AxiosResponse<IRefreshResponse> = await instance(config);

    console.log('token refreshed', new Date());

    if (access_token && refresh_token) {
      await saveToken('access_token', access_token);
      await saveToken('refresh_token', refresh_token);

      failedRequest.response.config.headers['Authorization'] = 'Bearer ' + access_token;

      return Promise.resolve();
    } else {
      console.error('Expected token');
    }
  } catch (e) {
    console.error(e);
  }

  return Promise.reject();
};

createAuthRefreshInterceptor(instance, refreshAuthLogic, { pauseInstanceWhileRefreshing: false });

instance.interceptors.request.use(async request => {
  request.headers['Authorization'] = `Bearer ${await retrieveToken('access_token')}`;
  return request;
});
```

This is the code I use to make an API call with a bad token:
```ts
await saveToken('access_token', 'badToken');

const config: AxiosRequestConfig = {
  method: 'get',
  [...]
};

try {
  console.log('Before API call', new Date());
  const response: AxiosResponse = await instance(config);
  console.log('After API call', new Date());
} catch (e) {
  console.error('API call failed', e);
}
```

With the official 3.1.0 on npm (which does not carry https://github.com/Flyrell/axios-auth-refresh/pull/121) the previous code does not work at all: it never gets past the following line:
```ts
    const {data: {access_token, refresh_token}}: AxiosResponse<IRefreshResponse> = await instance(config);
```

A workaround is to use another instance of axios:

```ts
    const {data: {access_token, refresh_token}}: AxiosResponse<IRefreshResponse> = await axios(config);
```

This PR (https://github.com/Flyrell/axios-auth-refresh/pull/163) fixes the issue, but there is still a very long (~32 seconds) delay between the failed (401) API call and the moment `refreshAuthLogic` gets run:

```
pauseInstanceWhileRefreshing: false

19:11:36 Before API call
19:12:08 Got 401, refreshing token +32
19:12:10 token refreshed +2
19:12:11 After API call +1
```

Ironically if I set `pauseInstanceWhileRefreshing` to `true` I get another 36 seconds delay between the moment the token got refreshed and the moment we retrieve the result from the second API call (the one with the good token):

```
pauseInstanceWhileRefreshing: true

19:08:36 Before API call
19:09:12 Got 401, refreshing token +36
19:09:14 token refreshed +2
19:09:50 After API call +36
```

Any idea how to fix this?

**Comment by darkbasic** [comment]
I forgot to specify I'm using react-native, not sure if matters.

**Comment by SudoPlz** [comment]
@darkbasic I'm using react-native too. This PR https://github.com/Flyrell/axios-auth-refresh/pull/121 fixes the issue.

I had to download the files locally to apply those changes (at least until that PR is merged and goes live)

**Comment by darkbasic** [comment]
@SudoPlz I'm already using this PR (almost, I made my own because it fails to compile: https://github.com/Flyrell/axios-auth-refresh/pull/163) and using it allows me to use the same axios instance. Unfortunately, even with this PR, there is a 32 seconds delay before `refreshAuthLogic` gets called.

**Comment by darkbasic** [comment]
I've found the issue and added a commit to my PR: https://github.com/Flyrell/axios-auth-refresh/pull/163/commits/9d143c386735c8e7470157d62a3a90c8a20abab7

**Comment by fblenkle** [comment]
Are there any plans for a new release with PR #121? Ran into the same problems and need a solution. The workaround with another instance is too complex in my app as I have multiple interceptors/config which would get lost.

**Comment by darkbasic** [comment]
@fblenkle #121 on its own is not enough, I've pushed a couple more fixes on top of it in PR #163

**Comment by Flyrell** [comment]
Merged in #163 with a fix in #166. Thanks for the contributions.
